### PR TITLE
HPCC-14286 Redis Plugin - add publish and subscribe functions

### PR DESCRIPTION
**Pull request by jamienoss**
Signed-off-by: James Noss james.noss@lexisnexis.com

### CODE REVIEW

**Comment by jamienoss** [comment]
@richardkchapman, @ghalliday please review.
I've been sitting on this whilst I find a way to add a test to the regression test. Thought I'd get the ball rolling on the review and get back to finding a suitable test - will discuss tomorrow.

File: plugins/redis/README.md [hunk 1]
```diff
@@ -85,6 +85,12 @@ Expire(CONST VARSTRING key, CONST VARSTRING options, UNSIGNED database = 0, UNSI
 INTEGER DBSize(CONST VARSTRING options, UNSIGNED database = 0, CONST VARSTRING password = '', UNSIGNED timeout = 1000)
 ```
 
+###PUB-SUB
+```
+EXPORT Publish(VARSTRING keyOrChannel, STRING message, INTEGER4 database = 0, BOOLEAN lockedKey = FALSE) := redis.Publish(keyOrChannel, message, options, database, password, timeout, lockedKey);
```
Comment (ghalliday) [inline plugins/redis/README.md]: timeout and paswsword aren't in the parameters.
Also, not consistent with the definitions above which are just the prototypes.

File: plugins/redis/redis.cpp [hunk 2]
```diff
@@ -560,37 +595,110 @@ template<class type> void Connection::get(ICodeContext * ctx, const char * key,
...
+#if(0)//Replicate a lingering SUBSCRIBE to test channel comparison test when reading message.
+    {
+    OwnedReply reply = Reply::createReply(redisCommand(context, "SUBSCRIBE oldChannel"));
+    assertOnErrorWithCmdMsg(reply->query(), "Test lingering SUBSCRIBE", "oldChannel");
+    }
+#endif
+
+    OwnedReply reply = Reply::createReply(redisCommand(context, "SUBSCRIBE %b", channel.str(), (size_t)channel.length()));
+    assertOnErrorWithCmdMsg(reply->query(), "SUBSCRIBE", channel.str());
+    if (reply->query()->type == REDIS_REPLY_ARRAY && strcmp("subscribe", reply->query()->element[0]->str) != 0 )
```
Comment (ghalliday) [inline plugins/redis/redis.cpp]: should this be

reply->query()->type != REDIS_REPLY_ARRAY || strcmp...

?

File: plugins/redis/redis.cpp [hunk 3]
```diff
@@ -560,37 +595,110 @@ template<class type> void Connection::get(ICodeContext * ctx, const char * key,
...
+        fail("SUBSCRIBE", "failed to register SUB", channel.str());
+
+    readReply(reply);
+    assertOnErrorWithCmdMsg(reply->query(), "SUBSCRIBE", channel.str());
+
+    if (reply->query()->type == REDIS_REPLY_ARRAY && strcmp("message", reply->query()->element[0]->str) == 0 && reply->query()->elements > 2)
```
Comment (ghalliday) [inline plugins/redis/redis.cpp]: Are the first two tests needed?

Comment (ghalliday) [inline plugins/redis/redis.cpp]: Ignore that.  I forgot readReply modifies its argument.

**Comment by ghalliday** [comment]
@jamienoss a couple of minor comments.  Also, the readme doesn't say how to use the two new functions - e.g., what does subscribe actually do?

File: plugins/redis/README.md (hunk 1 above)
Comment (jamienoss) [inline plugins/redis/README.md]: Good spot, copied the wrong section from the lib.

File: plugins/redis/redis.cpp (hunk 2 above)
Comment (jamienoss) [inline plugins/redis/redis.cpp]: Another good spot, thank you.

**Comment by jamienoss** [comment]
@ghalliday just testing a few changes that your comment regarding the logic triggered and a good shot on the bug causing Fridays seg fault.
Regarding the readme description of pub & sub, I haven't actually described any of the axillary functions e.g. clear, expire etc. I thought this best left to the redis doc. I will add a link to the relevant page though.

**Comment by jamienoss** [comment]
@ghalliday  back to you.
Note: the changes I mentioned about Fridays seg fault I have split out into
https://github.com/hpcc-systems/HPCC-Platform/pull/7877/files
https://github.com/hpcc-systems/HPCC-Platform/pull/7879/files

File: plugins/redis/redis.cpp [hunk 4]
```diff
@@ -103,14 +107,14 @@ private :
...
-    Connection(ICodeContext * ctx, const char * options, int database, const char * password, unsigned _timeout);
+    Connection(ICodeContext * ctx, const char * _options, int database, const char * password, unsigned _timeout);
     Connection(ICodeContext * ctx, const char * _options, const char * _ip, int _port, unsigned _serverIpPortPasswordHash, int _database, const char * password, unsigned _timeout);
     ~Connection()
     {
         if (context)
             redisFree(context);
     }
-    static Connection * createConnection(ICodeContext * ctx, const char * options, int database, const char * password, unsigned _timeout);
+    static Connection * createConnection(ICodeContext * ctx, Connection * & _chachedConnection, const char * options, int database, const char * password, unsigned _timeout, bool sub = false);
```
Comment (ghalliday) [inline plugins/redis/redis.cpp]: trivial: typo chached->cached

File: plugins/redis/redis.cpp [hunk 5]
```diff
@@ -330,11 +352,17 @@ void Connection::parseOptions(ICodeContext * ctx, const char * _options)
...
 }
-void Connection::resetContextErr()
+void Connection::reset()
 {
     if (context)
         context->err = REDIS_OK;
 }
+void SubConnection::reset()
+{
+    Connection::reset();
```
Comment (ghalliday) [inline plugins/redis/redis.cpp]: minor; Should this call be after the unsubscribe command?

**Comment by ghalliday** [comment]
@jamienoss looks good to merge.  A couple of minor questions you might want to reply to first.

File: plugins/redis/redis.cpp (hunk 4 above)
Comment (jamienoss) [inline plugins/redis/redis.cpp]: Cheers, changed.

File: plugins/redis/redis.cpp (hunk 5 above)
Comment (jamienoss) [inline plugins/redis/redis.cpp]: Yes, thanks again.

**Comment by jamienoss** [comment]
@ghalliday changes made and rebased.

**Pull request merged**